[DO NOT MERGE] Updated library mode README (#538)
NVIDIA/nv-ingest: README.md

## NVIDIA-Ingest: Multi-modal data extraction

> NVIDIA-Ingest is a scalable, performance-oriented content and metadata extraction SDK for a variety of input formats. NV-Ingest includes support for parsing PDFs, textfiles, Microsoft Word & PowerPoint documents, plain images, and audio files. NV-Ingest uses specialized NVIDIA NIMs (self-hosted microservices, or hosted on build.nvidia.com) to find, contextualize, and extract text, tables, charts and unstructured images that you can use in downstream generative applications.

Reviewer: ```suggestion
NVIDIA-Ingest is a scalable, performance-oriented content and metadata extraction SDK for a variety of input formats. NV-Ingest includes support for parsing PDFs, text files, Microsoft Word and PowerPoint documents, plain images, and audio files. NV-Ingest uses specialized NVIDIA NIMs (self-hosted microservices, or hosted on build.nvidia.com) to find, contextualize, and extract text, tables, charts, and unstructured images that you can use in downstream generative applications.
```

## What NVIDIA-Ingest Is ✔️

NV-Ingest is a library and microservice service that does the following:

> - Accept a job specification, containing a document payload, and a set of ingestion tasks to perform on that payload.

Reviewer: ```suggestion
- Accept a job specification that contains a document payload and a set of ingestion tasks to perform on that payload.
```

> - Allow the results of a job to be retrieved. The result is a dictionary that contains a list of metadata describing objects extracted from the base document, and processing annotations and timing/trace data.

Reviewer: ```suggestion
- Store the result of each job to retrieve later. The result is a dictionary that contains a list of metadata describing objects extracted from the base document, and processing annotations and timing/trace data.
```

> - Support multiple methods of extraction for each document type to balance trade-offs between throughput and accuracy. For example, for .pdf documents, we support extraction through pdfium, [nemoretriever-parse](https://build.nvidia.com/nvidia/nemoretriever-parse), Unstructured.io, and Adobe Content Extraction Services.

Reviewer: ```suggestion
- Support multiple methods of extraction for each document type to balance trade-offs between throughput and accuracy. For example, for .pdf documents nv-ingest supports extraction through pdfium, [nemoretriever-parse](https://build.nvidia.com/nvidia/nemoretriever-parse), Unstructured.io, and Adobe Content Extraction Services.
```

> - Support various types of pre- and post- processing operations, including text splitting and chunking, transform and filtering, embedding generation, and image offloading to storage.

Reviewer: ```suggestion
- Support various types of before and after processing operations, including text splitting and chunking, transform and filtering, embedding generation, and image offloading to storage.
```

- `tiff`
- `txt`

## Prerequisites

For production-level performance and scalability, we recommend that you deploy the pipeline and supporting NIMs by using Docker Compose or Kubernetes ([helm charts](helm)). For more information, refer to [prerequisites](https://docs.nvidia.com/nv-ingest/user-guide/getting-started/prerequisites).

## Library Mode Quickstart

For small-scale workloads, such as workloads of fewer than 100 PDFs, you can use library mode setup. Library mode set up depends on NIMs that are already self-hosted, or, by default, NIMs that are hosted on build.nvidia.com.

Library mode deployment of nv-ingest requires:

- Linux operating systems (Ubuntu 22.04 or later recommended)
- [Conda Python environment and package manager](https://github.com/conda-forge/miniforge)

> - Python 3.10

Reviewer: ```suggestion
- [Python version 3.10](https://www.python.org/downloads/release/python-3100/)
```

For production level performance and scalability, we recommend deploying the pipeline and supporting NIMs via Docker Compose or Kubernetes (via the [provided helm charts](helm). See the hardware and software requirments [here](https://docs.nvidia.com/nv-ingest/user-guide/getting-started/prerequisites).

## Library Mode Quickstart

For small-scale workloads, such as workloads of fewer than 100 PDFs, you can use library mode setup. Library mode set up depends on NIMs that are already self-hosted, or, by default, NIMs that are hosted on build.nvidia.com.

Library mode deployment of nv-ingest requires:

- Linux operating systems (Ubuntu 22.04 or later recommended)
- [Conda Python environment and package manager](https://github.com/conda-forge/miniforge)
- Python 3.10

### Step 1: Preparing Your Environment
```
conda create -y --name nvingest python=3.10
conda activate nvingest
conda install -y -c rapidsai -c rapidsai-nightly -c conda-forge -c nvidia nvidia/label/dev::nv_ingest nvidia/label/dev::nv_ingest_client nvidia/label/dev::nv_ingest_api
pip install opencv-python llama-index-embeddings-nvidia pymilvus 'pymilvus[bulk_writer, model]' milvus-lite dotenv ffmpeg nvidia-riva-client

> # Temporary workaround: remove after conda package fixes in place

Reviewer: ???

git clone https://github.com/nvidia/nv-ingest
cd nv-ingest
pip install -e ./api
pip install -e ./client
```

2. Create a .env file that contains your NVIDIA Build API key. For more information, refer to [Environment Configuration Variables](docs/docs/extraction/environment-config.md).

```
NVIDIA_BUILD_API_KEY=nvapi-...
NVIDIA_API_KEY=nvapi-...

> # For nvidians using nvdev endpoints:

Author: Yes. This PR will be amended once the unreleased NIMs are public on build.

# After release, public build endpoints are builtin to library setup
PADDLE_HTTP_ENDPOINT=https://ai.api.nvidia.com/v1/nvdev/cv/baidu/paddleocr
PADDLE_INFER_PROTOCOL=http

YOLOX_HTTP_ENDPOINT=https://ai.api.nvidia.com/v1/cv/nvdev/nvidia/nemoretriever-page-elements-v2
YOLOX_INFER_PROTOCOL=http

YOLOX_GRAPHIC_ELEMENTS_HTTP_ENDPOINT=https://ai.api.nvidia.com/v1/cv/nvdev/nvidia/nemoretriever-graphic-elements-v1
YOLOX_GRAPHIC_ELEMENTS_INFER_PROTOCOL=http

YOLOX_TABLE_STRUCTURE_HTTP_ENDPOINT=https://ai.api.nvidia.com/v1/cv/nvdev/nvidia/nemoretriever-table-structure-v1
YOLOX_TABLE_STRUCTURE_INFER_PROTOCOL=http

EMBEDDING_NIM_ENDPOINT=https://integrate.api.nvidia.com/v1
EMBEDDING_NIM_MODEL_NAME=nvdev/nvidia/llama-3.2-nv-embedqa-1b-v2

VLM_CAPTION_ENDPOINT=https://ai.api.nvidia.com/v1/gr/nvdev/meta/llama-3.2-11b-vision-instruct/chat/completions
VLM_CAPTION_MODEL_NAME=meta/llama-3.2-11b-vision-instruct

# NemoRetriever Parse:
NEMORETRIEVER_PARSE_HTTP_ENDPOINT=https://ai.api.nvidia.com/v1/cv/nvdev/nvidia/nemoretriever-parse
NEMORETRIEVER_PARSE_INFER_PROTOCOL=http
```

> ### Step 2: Ingesting Documents

Reviewer: ```suggestion
### Step 2: Ingest Documents
```

from nv_ingest_client.client import Ingestor, NvIngestClient
from nv_ingest_client.message_clients.simple.simple_client import SimpleClient
from nv_ingest.util.pipeline.pipeline_runners import PipelineCreationSchema
from nv_ingest_client.util.process_json_files import ingest_json_results_to_blob

# Start the pipeline subprocess for library mode                       
config = PipelineCreationSchema()                                                  
print(config)

pipeline_process = start_pipeline_subprocess(config)         
                                                                          
client = NvIngestClient(                                                                          
    message_client_allocator=SimpleClient,                                           
    message_client_port=7671,                                                                
    message_client_hostname="localhost"         
)                                                                  
                                            
# Note: gpu_cagra accelerated indexing is not yet available in milvus-lite

> # Providing a filename for milvus_uri will use milvus-lite

Reviewer: ```suggestion
# Provide a filename for milvus_uri to use milvus-lite
```

milvus_uri = "milvus.db"                
collection_name = "test"      
sparse=False

# do content extraction from files                                
ingestor = (
    Ingestor(client=client)
    .files("data/multimodal_test.pdf")
    .extract(              
        extract_text=True,
        extract_tables=True,
        extract_charts=True,
        extract_images=True,
        paddle_output_format="markdown",
        text_depth="page"
    ).embed()
    .caption(           
        model_name="nvdev/meta/llama-3.2-11b-vision-instruct",
        endpoint_url="https://ai.api.nvidia.com/v1/gr/nvdev/meta/llama-3.2-11b-vision-instruct/chat/completions"
    )
    .vdb_upload(
        collection_name=collection_name,
        milvus_uri=milvus_uri,
        sparse=sparse,
        # for llama-3.2 embedder, use 1024 for e5-v5
        dense_dim=2048
    )
)

print("Starting ingestion..")
t0 = time.time()
results = ingestor.ingest()
t1 = time.time()
print(f"Time taken: {t1-t0} seconds")

# results blob is directly inspectable
print(ingest_json_results_to_blob(results[0]))
```

> You can see the extracted text representing the content of the ingested test document:

Reviewer: ```suggestion
You can see the extracted text that represents the content of the ingested test document by using the following code.
```

```shell
...
Chart 2
This chart shows some average frequency ranges for speaker drivers.
Conclusion
This is the conclusion of the document. It has some more placeholder text, but the most 
important thing is that this is the conclusion. As we end this document, we should have 
been able to extract 2 tables, 2 charts, and some text including 3 bullet points.
image_caption:[]

Prompt: Using the following content: TestingDocument
A sample document with headings and placeholder text
Introduction
This is a placeholder document that can be used for any purpose. It contains some 
headings and some placeholder text to fill the space. The text is not important and contains 
no real value, but it is useful for testing. Below, we will have some simple tables and charts 
that we can use to confirm Ingest is working as expected.
Table 1
This table describes some animals, and some activities they might be doing in specific 
locations.
Animal Activity Place
Gira@e Driving a car At the beach
Lion Putting on sunscreen At the park
Cat Jumping onto a laptop In a home o@ice
Dog Chasing a squirrel In the front yard
Chart 1
This chart shows some gadgets, and some very fictitious costs.
```

> ### Step 3: Query Over Ingested Content

Reviewer: ```suggestion
### Step 3: Query Ingested Content
```

> Next you can query for relevant snippets of the ingested content and use it with LLMs to generate answers:

Reviewer: ```suggestion
To query for relevant snippets of the ingested content, and use it with LLMs to generate answers, use the following code.
```
